DirectProgramming/C++SYCL_FPGA/Tutorials/Tools/use_library/part1_native/src/use_native.cpp

```diff
@@ -0,0 +1,107 @@
+class IDPipeA;
+using InputPipeA = sycl::ext::intel::experimental::pipe<IDPipeA, unsigned>;
+class IDPipeB;
+using InputPipeB = sycl::ext::intel::experimental::pipe<IDPipeB, unsigned>;
+class IDPipeC;
+using OutputPipeC = sycl::ext::intel::experimental::pipe<IDPipeC, unsigned long>;
+
+// This kernel computes multiplier result by using the C++ '*' operator
+template <typename PipeIn1, typename PipeIn2, typename PipeOut>
+struct NativeMult27x27 {
+
+  // use a streaming pipelined invocation interface to minimize hardware
+  // overhead
+  auto get(sycl::ext::oneapi::experimental::properties_tag) {
+    return sycl::ext::oneapi::experimental::properties{
+        sycl::ext::intel::experimental::streaming_interface_accept_downstream_stall, 
+        sycl::ext::intel::experimental::pipelined<1>};
+  }
+  
+   void operator()() const {
+    MyInt27 a_val = PipeIn1::read();
+    MyInt27 b_val = PipeIn2::read();
+    MyInt54 res =(MyInt54)a_val * b_val;
+    PipeOut::write(res);
+  }
+};
+
+// This kernel compute result by performing the basic multipler soft logic
+int main() {
+  unsigned long result_native = 0;
+  unsigned kA = 134217727;  // 0x7FFFFFF is the largest possible ac_int<27, false>.
+  unsigned kB = 100;
+
+  // Select the FPGA emulator (CPU), FPGA simulator, or FPGA device
+#if FPGA_SIMULATOR
+  auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+  auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+  auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+  try {
+    sycl::queue q(selector, fpga_tools::exception_handler);
+
+    auto device = q.get_device();
+
+    std::cout << "Running on device: "
+              << device.get_info<sycl::info::device::name>().c_str()
+              << std::endl;
+    {
+      // write data to host-to-device pipes
+      InputPipeA::write(q, kA);
+      InputPipeB::write(q, kB);
+      // launch kernel that infers a multiplier automatically
+      q.single_task<KernelCompute>(NativeMult27x27<InputPipeA,InputPipeB,OutputPipeC>{}).wait();
+      // read data from device-to-host pipe
+      result_native = OutputPipeC::read(q);
+    }
+
+  } catch (sycl::exception const &e) {
+    // Catches exceptions in the host code
+    std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.code().value() == CL_DEVICE_NOT_FOUND) {
+      std::cerr << "If you are targeting an FPGA, please ensure that your "
+                   "system has a correctly configured FPGA board.\n";
+      std::cerr << "Run sys_check in the oneAPI root directory to verify.\n";
+      std::cerr << "If you are targeting the FPGA emulator, compile with "
+                   "-DFPGA_EMULATOR.\n";
+    }
+    std::terminate();
+  }
+
+  // Check the results
+  unsigned long expected_result = (unsigned long) kA * kB;
```

Comment:
replace unsigned long with uint64_t
replace unsigned with uint32_t
```suggestion
  uint64_t expected_result = (uint64_t) kA * kB;
```